File: python-algorithm/ds_3_hash/L36-m.py
# -*- coding: utf-8 -*-


# 36. 有效的数独

# (i,j)属于某个宫格的计算公式为 : j//3 + (i//3) * 3
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        # 分别记录每行、每列、每个格内部是否出现数字1~9
        row = [[0 for i in range(10)] for i in range(9)]
        col = [[0 for i in range(10)] for i in range(9)]
        box = [[0 for i in range(10)] for i in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] == '.':
                    continue
                cur_num = int(board[i][j])
                # 判断是否出现过
                if row[i][cur_num] == 1:
                    return False
                if col[j][cur_num] == 1:
                    return False
                if box[j // 3 + (i // 3) * 3][cur_num] == 1:
                    return False
                row[i][cur_num] = 1
                col[j][cur_num] = 1
                box[j // 3 + (i // 3) * 3][cur_num] = 1
        return True

File: python-algorithm/ds_3_hash/L36-m.cpp

#include <iostream>
#include <unordered_map>
#include <vector>
#include <cstring>
#include <cassert>

using namespace std;
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        int row[9][10] = {0};// 存储每一行的每个数是否出现过
        // 整个board有9行，第二维的维数10是为了让下标有9，和数独中的数字9对应。
        int col[9][10] = {0};// 存储每一列的每个数是否出现过
        int box[9][10] = {0};// 存储每一个box的每个数是否出现过
        for(int i=0; i<9; i++){
            for(int j = 0; j<9; j++){
                // 遍历到第i行第j列的那个数,我们要判断这个数在其所在的行有没有出现过，
                // 同时判断这个数在其所在的列有没有出现过
                // 同时判断这个数在其所在的box中有没有出现过
                if(board[i][j] == '.') continue;
                int curNumber = board[i][j]-'0';
                if(row[i][curNumber]) return false;
                if(col[j][curNumber]) return false;
                if(box[j/3 + (i/3)*3][curNumber]) return false;

                row[i][curNumber] = 1;// 之前都没出现过，现在出现了，就给它置为1，下次再遇见就能够直接返回false了。
                col[j][curNumber] = 1;
                box[j/3 + (i/3)*3][curNumber] = 1;
            }
        }
        return true;
    }
};